### DISCUSSION TITLE
How to define arbitrary mass fractions for species in air5 gas mixture in Fortran?

### DISCUSSION THREAD

**Discussion by peyvanahmad** [mutationpp/Mutationpp#198] [Q&A]
I am completely new to mutation++. I am trying to couple mutation++ to my Fortran code. I have a couple of issues here. I initialize the "air5" gas mixture with "EquilTP" and set the state as "mpp_set_state(300.0, 101325.0, 1)" and calculate the mole and mass fractions of the species as:
           1 N          Mole frac=   3.9307751032005825E-080   Mass frac=   1.8348664984119937E-080
           2 O         Mole frac=   3.4974094164797889E-041    Mass frac=   1.8648358906364791E-041
           3 NO       Mole frac=   3.1835986981865580E-016   Mass frac=   3.1835986981865881E-016
           4 N2        Mole frac=  0.50000000000006994           Mass frac=  0.46679508499945316     
           5 O2        Mole frac=  0.49999999999992961           Mass frac=  0.53320491500054645 
For air at that temperature and pressure, I know the mole fraction of N2 must be around 0.8 and for O2 must be around 0.2. What is happening here? Is the definition of equilibrium here different from what we have for a dry air at atmospheric pressure? 

My other question is that, what if I want to define an arbitrary mass or mole fractions for species in a gas mixture? How can I define that in Fortran? I know we can use the "addComposition()" function in c++ to define the composition mass or mole fractions but how can we do that in Fortran?

**Comment by capriatim** [comment]
Hi,

can you post the mixture xml file? I think you should define the elemental composition there. Something like:
    <element_compositions default="air">
        <composition name="air"> N:0.79, O: 0.21 </composition>
    </element_compositions>

For the secondo question, if you do not want to run under LTE assumption, you can set the state_model to ChemNonEq1T and specify the composition of each species in the set state.

    /**
     * Sets the current state of the mixture.  Variable sets can be the
     * following:
     *   0: conserved variables (species densities, total energy density)
     *   1: primitive set 1 (species densities, mixture temperature)
     *   2: primitive set 2 (species mass fractions, {P, T} array)
     */
